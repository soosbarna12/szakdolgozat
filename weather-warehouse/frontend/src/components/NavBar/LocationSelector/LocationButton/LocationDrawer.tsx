import DeleteIcon from '@mui/icons-material/Delete';
import PlaceOutlinedIcon from '@mui/icons-material/PlaceOutlined';
import { Box, Drawer, IconButton, List, ListItem, ListItemText, Skeleton, Toolbar, Tooltip, Typography } from "@mui/material";
import dayjs from 'dayjs';
import React, { useContext } from "react";
import { HistoricalContext } from '../../../../contexts/HistoricalContext/HistoricalContext';
import { useDeleteLocationQuery } from '../../../../hooks/useDeleteLocationQuery';
import { useSavedLocationQuery } from '../../../../hooks/useSavedLocationsQuery';
import { StyledButton, StyledIconButton } from '../../../../stlyes/button.style';
import { LocationDrawerProps } from "./LocationDrawer.type";


export function LocationDrawer(props: Readonly<LocationDrawerProps>) {
  const { toggleLocationDrawer, open } = props;
  const { savedLocations, isLoading, refetch } = useSavedLocationQuery(open);
  const { mutateAsync: deleteLocation } = useDeleteLocationQuery();
  const { setHistoricalPageData, setLocation } = useContext(HistoricalContext);


  function handleDrawerClose() {
    toggleLocationDrawer(false);
  }

  function handleOpenSavedLocation(index: number) {
    if (!savedLocations) return;
    setHistoricalPageData(savedLocations[index].locationData);
    setLocation({ name: "", lat: 0, lon: 0 });
  }

  function handleDeleteLocation(userLocationID: number, event: React.MouseEvent<HTMLElement>) {
    event.stopPropagation();
    deleteLocation(userLocationID).then(() => { refetch() });
  }

  function renderLocations() {
    if (isLoading) {
      return [0, 1, 2].map((index) => (
        <ListItem key={index} data-testid="locationDrawerSkeleton">
          <Skeleton
            variant="rectangular"
            width="100%"
            height={100}
            sx={{ borderRadius: "20px" }}
          />
        </ListItem>
      ));
    }

    if (savedLocations && savedLocations?.length > 0) {
      return savedLocations.map((arrayElement, index) => {
        return (
          <ListItem key={index} disablePadding onClick={() => handleOpenSavedLocation(index)} data-testid="savedLocation">
            <StyledButton fullWidth variant="text" color="primary" sx={{ height: "100px", boxShadow: 4, margin: 1 }} >
              <ListItemText
                primary={arrayElement?.locationData?.at(-1)?.cityName}
                secondary={
                  <>
                    <span>Filtered Date: {arrayElement?.locationData?.at(-1)?.date}</span>
                    <span>Saved Date: {dayjs(arrayElement.dateSaved).format("YYYY-MM-DD")}</span>
                  </>}
              />
            </StyledButton>
            <IconButton
                edge="end"
                aria-label="delete"
                onClick={(event) => handleDeleteLocation(arrayElement.userLocationID, event)}
                sx={{ marginLeft: 1 }}
                data-testid="deleteButton"
              >
                <DeleteIcon />
              </IconButton>
          </ListItem >
        )
      });
    }

    return <Typography>No saved locations found.</Typography>
  }


  return (
    <Drawer
      open={open}
      onClose={handleDrawerClose}
      elevation={4}
      slotProps={{ backdrop: { style: { opacity: 1 } } }}>
      <Box sx={{ width: 300 }} paddingRight={"10px"} paddingLeft={"24px"} paddingTop={"8px"}>
        <List sx={{ padding: 0 }}>
          <Toolbar disableGutters>
            <Tooltip title="Close locations">
              <StyledIconButton
                onClick={handleDrawerClose}
                sx={{
                  padding: 0, margin: "10px", width: "48px", height: "48px",
                  borderRadius: "100%", boxShadow: 4, outline: 0
                }}>
                <PlaceOutlinedIcon sx={{ padding: 0, margin: 0, width: "32px", height: "32px" }} />
              </StyledIconButton>
            </Tooltip>
            <Typography variant="body1" textTransform={"uppercase"}>Saved locations</Typography>
          </Toolbar>
          {renderLocations()}
        </List>
      </Box >
    </Drawer>
  );
}
