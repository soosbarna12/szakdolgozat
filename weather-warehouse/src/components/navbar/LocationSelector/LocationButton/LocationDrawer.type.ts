export interface LocationDrawerProps {
	toggleLocationDrawer: (newOpen: boolean) => () => void;
	open: boolean;
}
