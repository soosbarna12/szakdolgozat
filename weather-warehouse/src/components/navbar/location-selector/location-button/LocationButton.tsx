import React from "react";


export function LocationButton() {
    return (
      <div>

      </div>
      
    );
  }