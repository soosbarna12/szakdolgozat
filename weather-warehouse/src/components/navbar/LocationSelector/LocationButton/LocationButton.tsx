import PlaceOutlinedIcon from '@mui/icons-material/PlaceOutlined';
import { Box, Tooltip } from "@mui/material";
import React from "react";
import { StyledIconButton } from '../../../../common.style';
import { LocationDrawer } from "./LocationDrawer";


export function LocationButton() {

  const [open, setOpen] = React.useState(false);

  function toggleLocationDrawer(newOpen: boolean) {
    setOpen(newOpen);
  };

  function handleDrawerOpen() {
    toggleLocationDrawer(true);
  }

  return (
    <Box sx={{ flexGrow: 0 }}>
      <Tooltip title="Open locations">
        <StyledIconButton
          onClick={handleDrawerOpen}
          sx={{ boxShadow: 4 }}>
          <PlaceOutlinedIcon sx={{ padding: 0, margin: 0, width: "32px", height: "32px" }} />
        </StyledIconButton>
      </Tooltip>
      <LocationDrawer toggleLocationDrawer={toggleLocationDrawer} open={open} />
    </Box >
  );
}
