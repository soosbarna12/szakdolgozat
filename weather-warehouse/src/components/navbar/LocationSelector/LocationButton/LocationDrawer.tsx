import PlaceOutlinedIcon from '@mui/icons-material/PlaceOutlined';
import { Box, Button, Drawer, IconButton, List, ListItem, ListItemText, Toolbar, Tooltip, Typography } from "@mui/material";
import React from "react";
import { LocationDrawerProps } from "./LocationDrawer.type";

const locations = ["Budapest", "New York", "Berlin"];

export function LocationDrawer(props: Readonly<LocationDrawerProps>) {
  const { toggleLocationDrawer, open } = props;

  function handleDrawerClose() {
    toggleLocationDrawer(false);
  }

  return (
    <Drawer
      open={open}
      onClose={handleDrawerClose}
      elevation={10}
      slotProps={{ backdrop: { style: { opacity: 0.25 } } }}>
      <Box sx={{ width: 300 }} paddingRight={"24px"} paddingLeft={"24px"}>
        <List>
          <Toolbar disableGutters>
            <Tooltip title="Close locations">
              <IconButton
                onClick={handleDrawerClose}
                color='secondary'
                sx={{ padding: 0, margin: "10px", width: "42px", height: "42px", borderRadius: "100%", boxShadow: 4, outline: 1 }}>
                <PlaceOutlinedIcon sx={{ padding: 0, margin: 0, width: "32px", height: "32px" }} />
              </IconButton>
            </Tooltip>
            <Typography>Saved locations</Typography>
          </Toolbar>

          {locations.map((text, index) => (
            <ListItem key={text} disablePadding onClick={handleDrawerClose}>
              <Button fullWidth variant="text" color="secondary"
                sx={{ boxShadow: 4, margin: 1 }}>
                <ListItemText primary={text} />
              </Button>
            </ListItem>
          ))}
        </List>
      </Box >
    </Drawer>

  );
}
