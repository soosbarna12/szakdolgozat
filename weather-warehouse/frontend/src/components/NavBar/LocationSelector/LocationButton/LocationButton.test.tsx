import '@testing-library/jest-dom';
import { fireEvent, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { renderWithQueryClient } from '../../../../utils/test/renderWithQueryClient';
import { LocationButton } from './LocationButton';

jest.mock('./LocationDrawer', () => ({
  LocationDrawer: ({ toggleLocationDrawer, open }: { toggleLocationDrawer: (open: boolean) => void; open: boolean }) => (
    <div data-testid="locationDrawer">
      <button data-testid="closeDrawerButton" onClick={() => toggleLocationDrawer(false)}>Close</button>
      {open && <span data-testid="drawerContent">Drawer is open</span>}
    </div>
  ),
}));

describe('NavBar/LocationSelector/LocationButton', () => {
  const renderComponent = () => {
    return renderWithQueryClient(<LocationButton />);
  };

  it('matches LocationButton component snapshot', () => {
    const component = renderComponent();
    expect(component).toMatchSnapshot();
  });

  it('renders the LocationButton and opens the LocationDrawer on click', async () => {
    renderComponent();

    // Verify the button is rendered
    const locationButton = screen.getByRole('button', { name: /open locations/i });
    expect(locationButton).toBeInTheDocument();

    // Click the button to open the drawer
    fireEvent.click(locationButton);

    // Verify the drawer is open
    await waitFor(() => {
      expect(screen.getByTestId('drawerContent')).toBeInTheDocument();
    });
  });

  it('closes the LocationDrawer when the close button is clicked', async () => {
    renderComponent();

    // Open the drawer
    const locationButton = screen.getByRole('button', { name: /open locations/i });
    fireEvent.click(locationButton);

    // Verify the drawer is open
    await waitFor(() => {
      expect(screen.getByTestId('drawerContent')).toBeInTheDocument();
    });

    // Click the close button
    const closeButton = screen.getByTestId('closeDrawerButton');
    fireEvent.click(closeButton);

    // Verify the drawer is closed
    await waitFor(() => {
      expect(screen.queryByTestId('drawerContent')).not.toBeInTheDocument();
    });
  });
});