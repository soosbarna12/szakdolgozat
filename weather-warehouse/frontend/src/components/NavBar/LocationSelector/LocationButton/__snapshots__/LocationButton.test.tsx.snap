// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar/LocationSelector/LocationButton matches LocationButton component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-2uchni"
      >
        <button
          aria-label="Open locations"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-7l956p-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          data-testid="locationButton"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1gtean2-MuiSvgIcon-root"
            data-testid="PlaceOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2m6-1.8C18 6.57 15.35 4 12 4s-6 2.57-6 6.2c0 2.34 1.95 5.44 6 9.14 4.05-3.7 6-6.8 6-9.14M12 2c4.2 0 8 3.22 8 8.2 0 3.32-2.67 7.25-8 11.8-5.33-4.55-8-8.48-8-11.8C4 5.22 7.8 2 12 2"
            />
          </svg>
        </button>
        <div
          data-testid="locationDrawer"
        >
          <button
            data-testid="closeDrawerButton"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-2uchni"
    >
      <button
        aria-label="Open locations"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-7l956p-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        data-testid="locationButton"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1gtean2-MuiSvgIcon-root"
          data-testid="PlaceOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 12c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2m6-1.8C18 6.57 15.35 4 12 4s-6 2.57-6 6.2c0 2.34 1.95 5.44 6 9.14 4.05-3.7 6-6.8 6-9.14M12 2c4.2 0 8 3.22 8 8.2 0 3.32-2.67 7.25-8 11.8-5.33-4.55-8-8.48-8-11.8C4 5.22 7.8 2 12 2"
          />
        </svg>
      </button>
      <div
        data-testid="locationDrawer"
      >
        <button
          data-testid="closeDrawerButton"
        >
          Close
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
